### PR TITLE
Add FrozenSet

### PR DESCRIPTION
**Pull request by CAD97**

### CODE REVIEW

**Comment by Manishearth** [comment]
It's unclear to me how this is useful, there's no point to having a `get(T) -> T::Target`; you already have the `T`.

An append-only hashset with no StableDeref requirements that does not expose the keys at all could work.

cc @eddyb

**Comment by CAD97** [comment]
`get` becomes a lot more beneficial if we had any of the three stdlib features `hash_raw_entry`, `map_get_key_value`, or `hash_set_entry`. Then you can do `get(&'a self, &'b str) -> &'a str)`.

Alternatively, we can use `hashbrown` directly, which does allow us to do this without waiting for stabilization of any of those.

**Comment by Manishearth** [comment]
As a cache upgrade, I see. Hmm.

File: src/set.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
+use std::cell::{Cell, UnsafeCell};
+use std::collections::hash_map::{Entry, HashMap};
+use std::hash::Hash;
+use std::iter::FromIterator;
+
+use stable_deref_trait::StableDeref;
+
+/// Append-only version of `std::collections::HashSet` where
+/// insertion does not require mutable access
+//  We don't use `FrozenSet<T, ()>`, as the `StableDeref` requirements are different
```
Comment (eddyb) [inline src/set.rs]: I think you meant `FrozenMap` here?

Comment (CAD97) [reply src/set.rs]: whoops, accidentally hit this one when I was going through fixing comments
```suggestion
//  We don't use `FrozenMap<T, ()>`, as the `StableDeref` requirements are different
```

**Pull request closed without merging**